libs/wingc/src/lsp/sync.rs

```diff
@@ -71,83 +98,99 @@ pub unsafe extern "C" fn wingc_on_did_change_text_document(ptr: u32, len: u32) {
 		eprintln!("Failed to parse 'did change' text document: {}", parse_string);
 	}
 }
+
 pub fn on_document_did_change(params: DidChangeTextDocumentParams) {
-	JSII_TYPES.with(|jsii_types| {
-		FILES.with(|files| {
-			let uri = params.text_document.uri;
-			let uri_path = uri.to_file_path().unwrap();
-			let path = uri_path.to_str().unwrap();
-
-			let result = partial_compile(
-				path,
-				params.content_changes[0].text.as_bytes(),
-				&mut jsii_types.borrow_mut(),
-			);
-			files.borrow_mut().insert(uri, result);
-		});
-	})
+	let DidChangeTextDocumentParams {
+		text_document,
+		content_changes,
+	} = params;
+
+	WING_TYPES.with(|wing_types| {
+		JSII_TYPES.with(|jsii_types| {
+			PROJECT_DATA.with(|project_data| {
+				let uri = text_document.uri;
+				let uri_path = uri.to_file_path().unwrap();
+				let source_text = content_changes.into_iter().next().unwrap().text;
+
+				partial_compile(
+					&uri_path,
+					source_text,
+					&mut wing_types.borrow_mut(),
+					&mut jsii_types.borrow_mut(),
+					&mut project_data.borrow_mut(),
+				);
+			});
+		})
+	});
 }
 
-/// Runs several phases of the wing compile on a file, including: parsing, type checking, and capturing
-fn partial_compile(source_file: &str, text: &[u8], jsii_types: &mut TypeSystem) -> FileData {
-	// Reset diagnostics before new compilation (`partial_compile` can be called multiple)
+/// Runs several phases of the wing compiler on a file, including: parsing, type checking, and lifting
+/// `ProjectData` is passed with results from previous compilations, and is updated with the results of this compilation.
+fn partial_compile(
+	source_path: &Path,
+	source_text: String,
+	mut types: &mut Types,
+	jsii_types: &mut TypeSystem,
+	project_data: &mut ProjectData,
+) {
```

Comment:
@yoav-steinberg That's mostly correct. Some of the parsing results are cached between executions of `partial_compile`, but everything after that (type-checking onwards) is re-done from scratch each time.

More specifically, we re-type-check "known" files every time. That includes all files the user has opened in the LSP, as well as any unopened Wing source files they depend on.

As far as what's reset and what's kept:
- `TypeSystem` is kept between compilations since it stores all of the information about JSII dependencies, and we don't want to reload / parse that every time.
- `ProjectData` (renamed from `FileData`) stores several pieces of information used by the LSP hooks, like the text, tree sitter tree, and AST for each source file. (Right now all of these pieces of data are just collected in a struct for convenience.) A lot of these that pertain to parsing are incrementally updated on each compilation.

`Types` is currently shared between compilations because our `Scope`'s have pointers to SymbolEnv's that we don't want to invalidate, but that might be addressed by https://github.com/winglang/wing/pull/3712, so we might be able to remove it as a parameter to this function. I'll add a TODO comment